### PR TITLE
blog introducing java controller framework

### PR DESCRIPTION
**Pull request by yue9944882**
as one of the maintainers of the offical java kubernetes client, we hope to share our recent work via kubernetes blog..

### CODE REVIEW

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
+
+---
+layout: blog
+title: "Cloud Native Java: Develop kubernetes controller application"
```
Comment (sftim) [inline content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]:
```suggestion
title: "Develop a Kubernetes controller in Java"
```

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+date: 2019-08-28
+slug: Cloud-Native-Java-Develop-Kubernetes-Controller-Application
+---
+
+**Authors:** Min Kim (Ant Financial), Tony Ado (Ant Financial)
+
+The [Kubernetes offical Java SDK](https://github.com/kubernetes-client/java) project
```
Comment (sftim) [inline content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]:
```suggestion
The official [Kubernetes Java SDK](https://github.com/kubernetes-client/java) project 
```

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+advanced workloads or systems.  
+
+## Overall
+
+Java is no doubt one of the most popular programming languages in the world but
+it's been difficult for a period time for those non-Golang developers to build up 
+their customized controller/operator due to the lack of library resources in the 
+community. In the world of Golang, there're already some excellent controller 
+frameworks, for example, [controller runtime](https://github.com/kubernetes-sigs/controller-runtime),
+[operator SDK](https://github.com/operator-framework/operator-sdk) and all these
```
Comment (sftim) [inline content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]:
```suggestion
[operator SDK](https://github.com/operator-framework/operator-sdk). These
```

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+existing Golang frameworks are relying on the various utilities from the 
+[Kubernetes Golang SDK](https://github.com/kubernetes/client-go) proven to 
+be stable over years. Driven by the emerging need of further integration into 
+the platform of Kubernetes, we not only ported many essential toolings from Golang
```
Comment (sftim) [inline content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]:
```suggestion
the platform of Kubernetes, we not only ported many essential toolings from the Golang
```

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+be stable over years. Driven by the emerging need of further integration into 
+the platform of Kubernetes, we not only ported many essential toolings from Golang 
+SDK into the kubernetes Java SDK including informers, work-queues, leader-elections, 
+etc. but also developed a controller-builder SDK which wires up everything into 
+a runnable controller without hiccups.
+
+
+## Backgrounds
+
+Before we dig further, we have to know why we probably need Java in Kubernetes:
```
Comment (sftim) [inline content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]:
```suggestion
Why use Java to implement Kubernetes tooling? You might pick Java for:
```

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+- __Integrating legacy enterprise Java systems__: Many companies have their legacy 
+systems or frameworks written in Java in favor of stability. We are not able to 
+move everything to Golang straightly.
```
Comment (sftim) [inline content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]: straightly? I'm not sure what you mean here.

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+successfully encapsulates the complex components inside controller into several 
+clean interfaces. With the help of Java Generics, we even move on a bit and simply 
+the encapsulation in a better way.
+
+As for more advanced usage, we can wrap multiple controllers into a controller-manager 
+or a leader-electing controller which helps deploying in HA setup. In a word, we can 
+basically find most of the equivalence implementations here from Golang SDK and
+more advanced features are under active development by us.
+
+## Planned for Future
```
Comment (sftim) [inline content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]:
```suggestion
## Future steps
```

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+or a leader-electing controller which helps deploying in HA setup. In a word, we can 
+basically find most of the equivalence implementations here from Golang SDK and
+more advanced features are under active development by us.
+
+## Planned for Future 
+
+The community behind the official Kubernetes Java SDK project will be focusing on
+providing more useful utilities for developers who hope to program cloud native
+Java applications to extend Kubernetes. If you are interested in more details, 
+please check out our repo [kubernetes-client/java](https://github.com/kubernetes-client/java).
```
Comment (sftim) [inline content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]:
```suggestion
please look at [kubernetes-client/java](https://github.com/kubernetes-client/java) on GitHub.
```

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+Feel free to share also your feedback with us, through Issues or Slack.
```
Comment (sftim) [inline content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]: Will readers know what Slack is, and if so which channel in which workspace?

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md (hunk 1 above)
Comment (yue9944882) [reply content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]: revised the title

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md (hunk 3 above)
Comment (yue9944882) [reply content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]: updated

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md (hunk 2 above)
Comment (yue9944882) [reply content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]: updated

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md (hunk 4 above)
Comment (yue9944882) [reply content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]: update

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md (hunk 5 above)
Comment (yue9944882) [reply content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]: copy/pasted

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md (hunk 6 above)
Comment (yue9944882) [reply content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]: i replaced w/ "easily"

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md (hunk 7 above)
Comment (yue9944882) [reply content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]: updated

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md (hunk 8 above)
Comment (yue9944882) [reply content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]: updated

File: content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md (hunk 9 above)
Comment (yue9944882) [reply content/en/blog/_posts/2019-08-28-cloud-native-java-controller-sdk.md]: added link to the channel

**Comment by yue9944882** [comment]
@sftim thanks, lmk if more updates is required

**Comment by mrbobbytables** [comment]
From a technical standpoint -- lgtm 👍

**Comment by sftim** [comment]
I'm fine with this, I'll let the blog reviewers take this forward formally.
/uncc

**Comment by tengqm** [comment]
This one has been hanging too long. Approving based on tech correctness confirmation and positive feedback on wording.
/lgtm
/approve

**Comment by zacharysarah** [comment]
@tengqm Please don't merge blog content without approval from the subproject lead, @kbarnard10. 

The blog is an [official subproject](https://github.com/kubernetes/community/tree/master/sig-docs#kubernetes-blog) with its own reviewers and approvers.

To be clear: don't merge PRs in subprojects where you do not have explicit permissions.

**Comment by Colstuwjx** [comment]
Hi @zacharysarah 

I have been submitted [an issue](https://github.com/kubernetes/test-infra/issues/15527#issuecomment-562944517) from test-infra project, and confirmed that we could avoid such things happen again by setting up the subproject OWNERS with `no_parent_owners: true`, it could limit the such blog related operations ONLY could be approved by blog subproject OWNERS.

Thanks.

**Pull request merged**